source/common/upstream/upstream_impl.cc

```diff
@@ -622,38 +622,140 @@ ClusterInfoImpl::ResourceManagers::load(const envoy::api::v2::Cluster& config,
       runtime, runtime_prefix, max_connections, max_pending_requests, max_requests, max_retries)};
 }
 
-StaticClusterImpl::StaticClusterImpl(const envoy::api::v2::Cluster& cluster,
-                                     Runtime::Loader& runtime, Stats::Store& stats,
-                                     Ssl::ContextManager& ssl_context_manager, ClusterManager& cm,
-                                     bool added_via_api)
-    : ClusterImplBase(cluster, cm.bindConfig(), runtime, stats, ssl_context_manager, added_via_api),
-      initial_hosts_(new HostVector()) {
+PriorityStateManager::PriorityStateManager(ClusterImplBase& cluster,
+                                           const LocalInfo::LocalInfo& local_info)
+    : parent_(cluster), local_info_node_(local_info.node()), priority_state_(1) {}
 
-  for (const auto& host : cluster.hosts()) {
-    initial_hosts_->emplace_back(HostSharedPtr{new HostImpl(
-        info_, "", resolveProtoAddress(host), envoy::api::v2::core::Metadata::default_instance(), 1,
-        envoy::api::v2::core::Locality().default_instance(),
-        envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())});
+void PriorityStateManager::initializePerPriority(
+    const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint) {
+  const uint32_t priority = locality_lb_endpoint.priority();
+  if (priority_state_.size() <= priority) {
+    priority_state_.resize(priority + 1);
+  }
+  if (priority_state_[priority].first == nullptr) {
+    priority_state_[priority].first.reset(new HostVector());
+  }
+  if (locality_lb_endpoint.has_locality() && locality_lb_endpoint.has_load_balancing_weight()) {
+    priority_state_[priority].second[locality_lb_endpoint.locality()] =
```

Comment:
I think this line is untested.  I think this PR might benefit from some tests aimed at PriorityStateManager - bounds checking, making sure it handles ill formed updates gracefully and has comments and ASSERTs for cases we believe can't be hit) etc.   The decoupling of resizing and adding hosts with priorities, for example, makes me a bit concerned a future user might end up with out of bounds crashes so commenting up the functions could be helpful theer